VideoLocker/src/test/java/org/edx/mobile/test/AuthenticationTests.java

```diff
@@ -0,0 +1,135 @@
+        mockServer.setDispatcher(dispatcher);
+        super.setUp();
+    }
+
+    private String getBaseMockUrl() {
+        return "http://" + mockServer.getHostName() + ":" + mockServer.getPort();
+    }
+
+    @Override
+    protected JsonObject generateConfigProperties() throws IOException {
+        // Add the mock host url in the test config properties
+        JsonObject properties = super.generateConfigProperties();
+        properties.addProperty(API_HOST_URL, getBaseMockUrl());
+        return properties;
+    }
+
+    @Test
+    public void with401_causedByExpiredToken_refreshAccessToken() throws Exception {
+        // create a client with the authenticator
+        client = client.newBuilder()
+                .authenticator(new OauthRefreshTokenAuthenticator(context))
+                .build();
+
+        // Build a new dummy request
+        Request request = new Request.Builder()
+                .url(mockServer.url("/dummy/endpoint/"))
+                .header("Authorization", "expired_token")
+                .build();
+
+        // Make request
+        Response response = client.newCall(request).execute();
+        assertEquals(200, response.code());
+        assertEquals("Bearer dummy", response.request().header("Authorization"));
+
+        RecordedRequest expiredRequest = mockServer.takeRequest();
+        assertEquals("/dummy/endpoint/", expiredRequest.getPath());
+        assertEquals("expired_token", expiredRequest.getHeader("Authorization"));
+
+        RecordedRequest refreshTokenRequest = mockServer.takeRequest();
+        assertEquals("/oauth2/access_token/", refreshTokenRequest.getPath());
+        assertEquals("grant_type=refresh_token&client_id=4a6972d2b4d60a962465&refresh_token=dummy_refresh_token",
+                refreshTokenRequest.getBody().readUtf8());
+
+        RecordedRequest refreshedRequest = mockServer.takeRequest();
+        assertEquals("/dummy/endpoint/", refreshedRequest.getPath());
+        assertEquals("Bearer dummy", refreshedRequest.getHeader("Authorization"));
+    }
+
+    @Test
+    public void with401_notCausedByExpiredToken_doNotRefreshAccessToken() throws Exception {
+        client = client.newBuilder()
+                .authenticator(new OauthRefreshTokenAuthenticator(context))
+                .build();
+
+        Request request = new Request.Builder()
+                .url(mockServer.url("/dummy/endpoint/"))
+                .header("Authorization", "401_not_caused_by_expired_token")
+                .build();
+
+        Response response = client.newCall(request).execute();
+        assertEquals(401, response.code());
+        assertEquals("401_not_caused_by_expired_token", response.request().header("Authorization"));
+    }
+
+    final Dispatcher dispatcher = new Dispatcher() {
+
+        @Override
+        public MockResponse dispatch(RecordedRequest request) throws InterruptedException {
```

Comment:
Should we have a bunch of separate dispatchers? Or just enqueue responses? It makes it kind of hard to follow what the expected output is going to be when you make the request in the test